src/shogun/kernel/normalizer/SqrtDiagKernelNormalizer.h

```diff
@@ -121,45 +117,35 @@ class SqrtDiagKernelNormalizer : public KernelNormalizer
 		 * alloc and compute the vector containing the square root of the
 		 * diagonal elements of this kernel.
 		 */
-		bool alloc_and_compute_diag(Kernel* k, float64_t* &v, int32_t num) const
+		SGVector<float64_t> alloc_and_compute_diag(Kernel* k, int32_t num) const
 		{
-			SG_FREE(v);
-			v=SG_MALLOC(float64_t, num);
+			SGVector<float64_t> v(num);
+			std::function<float64_t(const int32_t)> func;
 
-			for (int32_t i=0; i<num; i++)
+			if (auto* cwk = dynamic_cast<CommWordStringKernel*>(k))
 			{
-				if (k->get_kernel_type() == K_COMMWORDSTRING)
-				{
-					auto cwk = k->as<CommWordStringKernel>();
-					if (use_optimized_diagonal_computation)
-						v[i]=sqrt(cwk->compute_diag(i));
-					else
-						v[i]=sqrt(cwk->compute_helper(i,i, true));
-				}
+				if (use_optimized_diagonal_computation)
+					func = [&cwk] (const int32_t i) {return std::sqrt(cwk->compute_diag(i));};
 				else
-					v[i]=sqrt(k->compute(i,i));
-
-				if (v[i]==0.0)
-					v[i]=1e-16; /* avoid divide by zero exception */
+					func = [&cwk] (const int32_t i) { return std::sqrt(cwk->compute_helper(i,i, true)); };
 			}
+			else 
+				func = [&k] (const int32_t i) { return std::sqrt(k->compute(i,i));};
 
-			return (v!=NULL);
+			for (int32_t i=0; i<num; i++)
+				v[i]= std::max(func(i), 1e-16);
```

Comment:
@karlnapf is it safe to assume that a kernel value is always positive? Because otherwise I need to change this back. This is just making sure that the value is never exactly zero.

Reply:
nope ... kernels can be negative pointwise. The kernel matrix needs to be psd for any input ...